kpi/deployment_backends/mock_backend.py

```diff
@@ -201,6 +206,23 @@ def get_submissions(self, requesting_user_id,
 
         return submissions
 
+    def duplicate_submission(self, requesting_user_id, instance_id, **kwargs):
+        all_submissions = self.asset._deployment_data['submissions']
+        submission = next(
+            filter(lambda sub: sub['_id'] == instance_id, all_submissions)
+        )
+        next_id = max([sub['_id'] for sub in all_submissions]) + 1
+        duplicate_submission = copy.copy(submission)
+        updated_time = datetime.now(tz=pytz.UTC).isoformat('T', 'milliseconds')
+        updated_fields = {
+                '_id': next_id,
+                'start': updated_time,
+                'end': updated_time,
+                'instanceID': f'uuid:{uuid.uuid4()}'
+                }
+
+        return {**duplicate_submission, **updated_fields}
```

Comment:
Now that `KobocatDeploymentBackend.duplicate_submission()` actually returns a json/dict response, the mocked `MockDeploymentBackend.duplicate_submission()` is now actually pretty close. But thanks for spotting that anyway 🕵️‍♂️

Reply:
This in-place create-and-update is pretty clever: `{**duplicate_submission, **updated_fields}` :ok_hand:
I didn't know this kind of thing was allowed(!)
```
>>> {'a': 1, 'b': 2, 'c': 3, 'a': '9000'}
{'a': '9000', 'b': 2, 'c': 3}
```

But :nerd_face: by doing this you render `copy.copy(submission)` unnecessary, since `{}` is also creating a new dictionary:
```
>>> orig_dict = dict(a=1, b=2, c=3, d=4)
>>> new_dict = {**orig_dict}
>>> id(orig_dict)
140347297415936
>>> id(new_dict)
140347297417792
```

On the other hand, `copy.copy()` is needed if you go with the more traditional(?) approach of calling `update()`:
```
>>> orig_dict = dict(a=1, b=2, c=3, d=4)
>>> new_dict = copy.copy(orig_dict)
>>> new_dict.update({'a': 1000, 'b': 2000})
>>> orig_dict
{'a': 1, 'b': 2, 'c': 3, 'd': 4}
>>> new_dict
{'a': 1000, 'b': 2000, 'c': 3, 'd': 4}
```

I also didn't know you could use this syntax, but you can:
```
>>> new_dict.update(c=3000, d=4000)
>>> new_dict
{'a': 1000, 'b': 2000, 'c': 3000, 'd': 4000}
```

Reply:
Very interesting 🤔 good spot with the redundant `copy.copy(submission)`. Which method of updating/merging dicts to you prefer going forward? In JS I would usually do something like `{...dict_1, ...dict_2}` — hence the `{**dict_1, **dict_2}`